agent/src/action.rs

```diff
@@ -0,0 +1,156 @@
+use std::{
+    sync::{Arc, Mutex},
+    time::Duration,
+};
+
+use bollard::exec::{CreateExecResults, StartExecResults};
+use futures_util::StreamExt;
+use log::info;
+use tokio::{spawn, sync::mpsc::UnboundedSender, time::sleep};
+use tonic::Status;
+
+use crate::{
+    container::{
+        create_exec, inspect_exec, launch_container, remove_container, start_exec, stop_container,
+    },
+    proto::{ActionResponseStream, ActionResult},
+};
+
+// TODO: manage all unwrap on lock on log_input and action_id
+
+pub async fn launch_action(
+    image_name: String,
+    commands: &mut Vec<String>,
+    log_input: Arc<Mutex<UnboundedSender<Result<ActionResponseStream, Status>>>>,
```

Comment:
I have trouble understanding the point of the `UnboundedSender` type (from the futures package)

Doc says it's
> The transmission end of an unbounded mpsc ("Multi-producer, single-consumer FIFO queue communication primitives") channel.

Could you kindly help me understand its role in this code?

Reply:
Yep, mpsc::unbounded_channel alow to send data in a UnboundedSender to collect it later in a UnboundedReceiver. 
In this function I am using it to send log that inform the scheduler the state of this action and also that expose container logs by using log_input.send() .

Reply:
Okay, we'll discuss this on the technical pov later ; thanks